### PR TITLE
Grunt and bower instead of assetic

### PR DESCRIPTION
**Pull request by tijsverkoyen**
# Using grunt

We are using grunt to handle a shitload of frontend related tasks, below you 
can find all stuff that is done. This is more a guide for you to understand 
what is happening.

## @todo

* [ ] shouldn't we include the css/js from outside the document root as these 
    will be "compiled" with assetic
* [ ] fix deployment
* [x] fix issue with bootstrap glyphicons-halflings
* [x] fix issue with jquery.min.map
* [ ] try to find solution for livereload when `grunt watch` isn't running
* [x] remove references to my local repo's
* [x] tag our bundles and update composer.json
* [x] fix issue with woff2-file

### Coffee

We will look into all bundles in the `src`-folder for all files in the 
`Resources/assets/coffee`-folder, these files will be copied to the 
`assets/coffee`-folder in the root directory. All files will be prefixed with 
the lowercased bundlename.

After that the files will be compiled into `web/assets/js`.

As this will generate seperate files for each file we will combine them through
assetic. See [How to use Assetic for asset management](http://symfony.com/doc/current/cookbook/assetic/asset_management.html)
for the full documentation but basically:

    {% javascripts
        'assets/js/frameworkcorebundle.framework.data.js'
        'assets/js/frameworkcorebundle.framework.locale.js'
        'assets/js/frameworkcorebundle.framework.form.js'
        'assets/js/frameworkcorebundle.framework.js'
        'assets/js/frameworkcorebundle.framework.form.search.js'
        'assets/js/frameworkcorebundle.app.js'
        output='js/app.js'
    %}
        <script src="{{ asset_url }}"></script>
    {% endjavascripts %}

### Fonts

We will look into all bundles in the `src`-folder for all the ttf and otf-files
in the `Resources/assets/fonts`-folder, these files will be copied to 
`assets/fonts`-folder in the root directory. When this is done all the other 
formats (eot, svg, ttf, woff) are created in the `web/assets/fonts`-folder.

You can link to the font-files with the compass-shortcurt font-url('filename') 
in your scss-files.

### Icons

We will look into all bundles in the `src`-folder for all the svg-files in the 
`Resources/assets/icons`-folder, these files will be compiled into an iconfont 
located in `web/assets/fonts`-folder in the root directory. A _icons.scss-file 
will be stored in the `assets/sass`-folder.

### Images

We will look into all bundles in the `src`-folder for all files in the 
`Resources/assets/images`-folder, these files will be copied to the 
`web/assets/images`-folder in the root directory. The folder structure you 
(optionally) created will be preserved. 

*Important*: make sure you don't have duplicate filenames as the files will be 
overwritten.

You can link to the font-files with the compass-shortcurt image-url('filename')`
in your scss-files.

When running `grunt build` the images will be optimized for web with the 
[grunt-contrib-imagemin-plugin](https://www.npmjs.com/package/grunt-contrib-imagemin).

You can use the `asset`-method in twig templates like below:

    

### JS

We will look into all bundles in the `src`-folder for all files in the 
`Resources/assets/js`-folder, these files will be copied to the 
`web/assets/js`-folder in the root directory. The folder structure you 
(optionally) created will be preserved.

You can use the `asset`-method in twig templates like below:

    

### SASS/SCSS

We will look into all bundles in the `src`-folder for all files in the 
`Resources/assets/sass`-folder, these files will be copied to the 
`assets/sass`-folder in the root directory. The folder structure you 
(optionally) created will be preserved.

After that the files will be compiled into `web/assets/css`. When running 
`grunt build` the files will be compiled as minified as possible, so without 
any comments.

When the sass/scss-files are compiled we use the 
[grunt-autoprefixer-plugin](https://www.npmjs.com/package/grunt-autoprefixer)

If you want you can combine multiple generated css files into one by using 
Assetic:

    {% stylesheets
        'assets/css/1.css'
        'assets/css/2.css'
        filter='cssrewrite'
        output='css/combined.css'
    %}
        <link rel="stylesheet" href="{{ asset_url }}" />
    {% endstylesheets %}

But in most cases you will just use

    {% stylesheets
        'assets/css/style.css'
        filter='cssrewrite'
        output='css/style.css'
    %}
        <link rel="stylesheet" href="{{ asset_url }}" />
    {% endstylesheets %}

### While developing

    grunt
    
We have implemented live-reload, so your changes will be reloaded in the 
browser. This will only happen in the dev-environment.

### Before launching your website

    grunt build
    
You don't have to bother if you are using deployment as we will handle it for 
you.

### CODE REVIEW

**Comment by tijsverkoyen** [comment]
Note that this is a work in progress

File: composer.json [hunk 1]
```diff
@@ -6,15 +6,30 @@
     "autoload": {
         "psr-0": { "": "src/" }
     },
+    "repositories": [
+        {
+            "type": "vcs",
+            "url" : "/Users/tijs/Sites/sumocoders/frameworkBundles/FrameworkErrorBundle"
```
Comment (WouterSioen) [inline composer.json]: These should be removed

File: src/SumoCoders/FrameworkCoreBundle/Menu/MenuBuilder.php [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+     */
+    private $eventDispatcher;
+
+    /**
+     * @var FactoryInterface
+     */
+    private $factory;
+
+    /**
+     * @param \Symfony\Component\EventDispatcher\EventDispatcherInterface $eventDispatcher
```
Comment (WouterSioen) [inline src/SumoCoders/FrameworkCoreBundle/Menu/MenuBuilder.php]: Can't you just use @param EventDispatcherInterface when the use statement is available?

File: composer.json (hunk 1 above)
Comment (tijsverkoyen) [inline composer.json]: Fixed in cfc72bebb55cda9f3b67b7b3154c50ea464d2fad

File: src/SumoCoders/FrameworkCoreBundle/Menu/MenuBuilder.php (hunk 2 above)
Comment (tijsverkoyen) [inline src/SumoCoders/FrameworkCoreBundle/Menu/MenuBuilder.php]: Fixed in 0180fc4edf040f22dddd6597c1f003138438ba71

File: Gruntfile.coffee [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+        'sync:js'
+        'sync:sass'
+      ]
+      syncImagesAndJs: [
+        'sync:images'
+        'sync:js'
+      ]
+      generateAssets: [
+        'generateFonts'
+        'generateCssForProduction'
```
Comment (WouterSioen) [inline Gruntfile.coffee]: Je gebruikt zowel "prod" als "Production" in deze file, ik zou voor een consistente benaming gaan (en imo is production beter).

File: Gruntfile.coffee [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+        ],
+        dest: '<%= webAssetsPath %>/js/'
+        rename: (dest, src) ->
+          dest + src.replace('.coffee', '.js')
+
+    # Fontgen configuration
+    fontgen:
+      all:
+        files: [
+          expand: true
```
Comment (WouterSioen) [inline Gruntfile.coffee]: Waarom bij all expand: true? Dat lijkt me iets dat enkel bij dev nodig is?

File: Gruntfile.coffee [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+  grunt.registerTask 'generateIconFonts', [
+    'clean:beforeGenerateIconFonts'
+    'webfont'
+  ]
+
+  # Generate the css-files
+  grunt.registerTask 'generateCssForDev', [
+    'compass'
+    'autoprefixer'
+    'clean:afterGenerateCssForProduction'
```
Comment (WouterSioen) [inline Gruntfile.coffee]: Je voert die zowel voor dev als voor productie uit. Is dit dan niet gewoon "afterGenerateCss"?

File: README.md [hunk 6]
```diff
@@ -1,11 +1,46 @@
 # Sumocoders Framework
 
+## Instalation
```
Comment (WouterSioen) [inline README.md]: Typo: Installation

File: TODO.md [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+* [ ] Logging, search for papertrail
+
+# Frontend
+
+* [ ] learn twig
+* [ ] assetic with grunt
+* [ ] upgrade HTML to bootstrap ...
+* [ ] buttons in header: correct width
+* [ ] Tijs: make Data.js work again
+* [ ] Tijs: make Locale.js work again
```
Comment (WouterSioen) [inline TODO.md]: Locale.js already works, could be removed from the checklist. (I think some other stuff could be ticked as well)

**Comment by WouterSioen** [comment]
When merging this, I'd squash some of the commits. I don't think it adds a lot of value to have 65 commits for this integration.

File: Gruntfile.coffee (hunk 4 above)
Comment (tijsverkoyen) [inline Gruntfile.coffee]: Die expand is nodig om de extra opties zoals  rename mogelijk te maken

File: Gruntfile.coffee (hunk 3 above)
Comment (tijsverkoyen) [inline Gruntfile.coffee]: Fixed in a7e727e426f7acb22598f28197aa1ae7a5f51518

File: Gruntfile.coffee (hunk 5 above)
Comment (tijsverkoyen) [inline Gruntfile.coffee]: Fixed in cf6de9555354fbadfb2c5a32fce603fa5040fc7c

File: README.md (hunk 6 above)
Comment (tijsverkoyen) [inline README.md]: Fixed in e8565351740a8a44826684a9c4af205056b060e2

File: TODO.md (hunk 7 above)
Comment (tijsverkoyen) [inline TODO.md]: fixed in d9828980b08002f6778c492727fdc71e6979ae40

**Comment by tijsverkoyen** [comment]
Squashed version of some commits in https://github.com/sumocoders/Framework/pull/5

**Pull request closed without merging**